### ISSUE TITLE
Mask RCNN tflite detection speed on android too slow

### ISSUE THREAD

**Issue by MeghaGhosh** [tensorflow/tensorflow#47973]
**System information**
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
  Android mobile
- TensorFlow installed from (source or binary):
   Binary
- TensorFlow version (use command below):
   2.4.0
- Python version:
  3.7

**Describe the current behavior**

We have converted Mask RCNN to TFLite using this link: 

https://wathek.medium.com/convert-mask-r-cnn-model-to-tflite-with-tensorflow-2-3-57160d3be18d and Tensorflow 2.4.0

We have used byte buffers for the input and output tensors and have the GPU delegate, multithreading enabled. However it is taking around 35 seconds for detection, which is really not feasible in terms of latency. 

**Describe the expected behavior**

We expected the inference time to be around 3-5 seconds. How can we reduce the time taken for detection? Are there any changes we can make to the model other than those done in the link above? Any help would be appreciated, thank you.

**Comment by Saduf2019** [comment]
@MeghaGhosh 

Please share simple standalone code such that we can replicate the issue reported or a colab gist with the error reported.

**Comment by abattery** [comment]
@srjoglekar246 could you take a look?

**Comment by srjoglekar246** [comment]
@MeghaGhosh it would be good to see the TFLite model. Can you share a link to the file?

**Comment by MeghaGhosh** [comment]
Hey, this is the google drive link to the tflite file:

https://drive.google.com/file/d/1S6T5GUHa9WrcK2WZa0DrHpD68w9jHnTX/view?usp=sharing

This is the code to run it on android:

imageBitmapForModel = BitmapFactory.decodeResource(con.getResources(), R.drawable.b1);
  if (imageBitmapForModel != null) {
      Log.e("Edgar", "Suces" + height + "width" + width);
  }

  ImageProcessor imageProcessor =
          new ImageProcessor.Builder()
                  .add(new ResizeOp(height, width, ResizeOp.ResizeMethod.NEAREST_NEIGHBOR))
                  .add(new ResizeWithCropOrPadOp(1024,1024))
                  .build();

  TensorImage tImage = new TensorImage(DataType.FLOAT32);
  tImage.load(imageBitmapForModel);
  tImage = imageProcessor.process(tImage);

  imgMeta = ByteBuffer.allocateDirect(1 * 18 * 4);
  imgMeta.order(ByteOrder.nativeOrder());

  anchors = ByteBuffer.allocateDirect(1 * 261888 * 4 * 4);
  anchors.order(ByteOrder.nativeOrder());

  imgMeta.rewind();

  for (int i = 0; i < 18; i++) {
      imgMeta.putFloat(0.0f);
  }
  anchors.rewind();
  for (int i = 0; i < 261888; i++) {
      for (int k = 0; k < 4; k++) {
          anchors.putFloat(0.0f);
      }

  }

  Object[] inputs = {tImage.getBuffer(), imgMeta, anchors};
  Map<Integer, Object> outputs = new HashMap<>();
  outputs.put(0, new float[1][1][1]);
  outputs.put(1, new float[1][1000][6][4]);
  outputs.put(2, new float[1][1000][6]);
  outputs.put(3, new float[1][100][6]);
  outputs.put(4, new float[1][100][28][28][6]);
  outputs.put(5, new float[1][1][4]);
  outputs.put(6, new float[1][1][2]);

  HashMap outputProbabilityBuffers = new HashMap<>();
  ByteBuffer output0 = ByteBuffer.allocateDirect(16000);
  output0.order(ByteOrder.nativeOrder());
  output0.rewind();
  ByteBuffer output1 = ByteBuffer.allocateDirect(1000 * 6 * 4 *4);
  output1.order(ByteOrder.nativeOrder());
  output1.rewind();
  ByteBuffer  output2 = ByteBuffer.allocateDirect(1000 * 6 * 4);
  output2.order(ByteOrder.nativeOrder());
  output2.rewind();
  ByteBuffer output3 = ByteBuffer.allocateDirect(100 * 6 * 4);
  output3.order(ByteOrder.nativeOrder());
  output3.rewind();

    float[][][] output3 = new float[1][100][6];
  ByteBuffer  output4 = ByteBuffer.allocateDirect(100 * 28 * 28 * 6 * 4);
  output4.order(ByteOrder.nativeOrder());
  output4.rewind();
  ByteBuffer  output5 = ByteBuffer.allocateDirect(4190208);
  output5.order(ByteOrder.nativeOrder());
  output5.rewind();
  ByteBuffer   output6 = ByteBuffer.allocateDirect(2095104);
  output6.order(ByteOrder.nativeOrder());
  output6.rewind();
  outputProbabilityBuffers.put(0, output0);
  outputProbabilityBuffers.put(1, output1);
  outputProbabilityBuffers.put(2, output2);
  outputProbabilityBuffers.put(3, output3);
  outputProbabilityBuffers.put(4, output4);
  outputProbabilityBuffers.put(5, output5);
  outputProbabilityBuffers.put(6, output6);


  Date date = Calendar.getInstance().getTime();
  DateFormat formatter = new SimpleDateFormat("dd/MM/yyyy-mm-ss");
  String today = formatter.format(date);
  Log.e("edgar Start", "" + today);

  tfLite.runForMultipleInputsOutputs(inputs, outputProbabilityBuffers);

      Date dates = Calendar.getInstance().getTime();
      DateFormat formatterdd = new SimpleDateFormat("dd/MM/yyyy-mm-ss");
      String todays = formatterdd.format(dates);
      Log.e("edgar End", "" + todays);

      final ArrayList<DetectionResult> recognitions = new ArrayList<>(NUM_DETECTIONS);
      return recognitions;

**Comment by srjoglekar246** [comment]
There are two issues here:

1. The model's input size, and consequently the internal tensors, are too large for existing mobile applications.
2. Acceleration (by GPU) isn't working because the input has dynamic batch size. Usually you can set a static batch size by providing `batch_size` to [keras.input]

So I would suggest, if possible, making the model *smaller* and *simpler*, hopefully with static batch sizes. Maybe you can distill the large MaskRCNN model to a smaller student model that has same Inputs & Outputs?

**Comment by MeghaGhosh** [comment]
Okay thanks, I'll look into this.

**Comment by MeghaGhosh** [comment]
Hi @srjoglekar246 I have removed the mask head, number of output tensors and reduced the image size (to 384*384), with a major compromise in accuracy to bring the inference time down to 6 seconds. (this is with multithreading)

This speed somewhat works for us however there are some problems. If you could shed some light as to how we can fix this:

The tflite output on android is different from the Keras h5 output. Reducing the image size has led to significantly fewer detections. For eg: we are getting these results in keras for a particular image:

[[[0.7725331  0.2580187  0.7878663  0.2675068  3.         0.97869164]
  [0.6697102  0.61200994 0.71017647 0.6631437  3.         0.95587164]
  [0.6538168  0.695219   0.68414    0.7051539  3.         0.8865816 ]
  [0.4635428  0.55197054 0.5702299  0.700203   2.         0.87292403]
  [0.6545733  0.67137647 0.6891192  0.70570266 3.         0.85790217]
  [0.6573738  0.62910855 0.7056983  0.7020159  3.         0.78753364]
  [0.         0.         0.         0.         0.         0.        ]
  [0.         0.         0.         0.         0.         0.        ]
  [0.         0.         0.         0.         0.         0.        ]....]


For the converted tflite model, we are getting only 2 outputs. This is not feasible as we have significantly compromised on accuracy. Also the results are quite weird as in the detections are repeated until the last one. We have faced this with each version of the model with different image sizes. For example, in our latest model the results are similar to this:

[[0.7725331  0.2580187  0.7878663  0.2675068  3.         0.97869164]
[0.7725331  0.2580187  0.7878663  0.2675068  3.         0.97869164]
[0.7725331  0.2580187  0.7878663  0.2675068  3.         0.97869164]
[0.7725331  0.2580187  0.7878663  0.2675068  3.         0.97869164]
[0.7725331  0.2580187  0.7878663  0.2675068  3.         0.97869164]
[0.7725331  0.2580187  0.7878663  0.2675068  3.         0.97869164]....
.........................................................................................................................................
..........................................................................................................................................
[0.7725331  0.2580187  0.7878663  0.2675068  3.         0.97869164]
[0.7725331  0.2580187  0.7878663  0.2675068  3.         0.97869164]
[0.6697102  0.61200994 0.71017647 0.6631437  3.         0.95587164]] 

Why are the detections being repeated? For eg: if there are 5 detections in total, the first 95 instances will be the same repeated values and then the last 5 instances will vary and will be our results.  I am sharing both the keras as well as tflite model here. Any help would be appreciated. The code is as mentioned above, we are using tensorflow lite support library to resize our image. The rest of the inputs match the keras inference.

This is the tflite model: https://drive.google.com/file/d/1ZUthHpTu1JtukSdTaHXaJzadh_iYdZF0/view?usp=sharing

This is the keras h5 model: https://drive.google.com/file/d/1pWrhJIplxkToi0hum9Ao-livFC45wz8b/view?usp=sharing

**Comment by junyiliu-1025** [comment]
Hi, I am new in tflite, and I got the same issue with you, Did you solve the problem?

**Comment by pjpratik** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The TFLite team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the TFLite space.

Thanks.
